File: 114.flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if (!root) {
            return;
        }
        if (root->left) {
            flatten(root->left);
        }
        if (root->right) {
            flatten(root->right);
        }
        TreeNode* right = root->right;
        root->right = root->left;
        root->left = nullptr;
        while (root->right) {
            root = root->right;
        }
        root->right = right;
    }
};

File: 114.flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        if not root:
            return
        if root.left:
            self.flatten(root.left)
        if root.right:
            self.flatten(root.right)
        left = root.left
        right = root.right
        root.right = root.left
        root.left = None
        while root.right:
            root = root.right
        root.right = right
